data/pools/transactionPool.go

```diff
@@ -160,45 +116,58 @@ func (pool *TransactionPool) Test(t transactions.SignedTxn) error {
 	pool.mu.RLock()
 	defer pool.mu.RUnlock()
 
-	_, _, _, err := pool.test(t)
-	return err
+	return pool.test(t)
 }
 
-func (pool *TransactionPool) test(t transactions.SignedTxn) (accountDeductions, bool, transactions.Txid, error) {
-	// check if we already have this transaction in the pool.
-	if _, has := pool.pendingTxns[t.ID()]; has {
-		return accountDeductions{}, false, transactions.Txid{}, errors.New("TransactionPool.test: transaction already in the pool")
+func (pool *TransactionPool) test(t transactions.SignedTxn) error {
+	if pool.pendingBlockEvaluator == nil {
+		return fmt.Errorf("TransactionPool.test: no pending block evaluator")
 	}
 
-	var minTransactionID transactions.Txid
-	isFull := len(pool.pendingTxns) >= pool.size
-
-	if isFull {
-		// transaction pool is full.
-		var minPriority transactions.TxnPriority
-		minTransactionID, minPriority = pool.txPriorityQueue.getMin()
+	tentativeRound := pool.pendingBlockEvaluator.Round() + pool.numPendingWholeBlocks
+	err := pool.pendingBlockEvaluator.TestTransaction(t, nil)
+	if err == ledger.ErrNoSpace {
+		tentativeRound++
+	} else if err != nil {
+		return err
+	}
 
-		if t.Priority().LessThan(minPriority.Mul(pool.exponentialPriorityGrowthFactor)) {
-			return accountDeductions{}, isFull, transactions.Txid{}, fmt.Errorf("TransactionPool.test: transaction pool is full and tx priority too low: min in pool %v vs. %v", minPriority, t.Priority())
+	if t.Txn.LastValid < tentativeRound {
+		return transactions.TxnDeadError{
+			Round:      tentativeRound,
+			FirstValid: t.Txn.FirstValid,
+			LastValid:  t.Txn.LastValid,
 		}
 	}
 
-	// check if we have committed the transaction recently already
-	committed, err := pool.ledger.Committed(t)
-	if err != nil {
-		return accountDeductions{}, isFull, transactions.Txid{}, fmt.Errorf("TransactionPool.test: failed to call Committed(): %v", err)
-	}
-	if committed {
-		return accountDeductions{}, isFull, transactions.Txid{}, fmt.Errorf("TransactionPool.test: transaction with ID %v has already been committed", t.ID())
+	// The baseline threshold fee per byte is 1, the smallest fee we can
+	// represent.  This amounts to a fee of 100 for a 100-byte txn, which
+	// is well below MinTxnFee (1000).  This means that, when the pool
+	// is not under load, the total MinFee dominates for small txns,
+	// but once the pool comes under load, the fee-per-byte will quickly
+	// come to dominate.
+	feePerByte := uint64(1)
+
+	// The threshold is multiplied by the feeThresholdMultiplier that
+	// tracks the load on the transaction pool over time.  If the pool
+	// is mostly idle, feeThresholdMultiplier will be 0, and all txns
+	// are accepted (assuming the BlockEvaluator approves them, which
+	// requires a flat MinTxnFee).
+	feePerByte = feePerByte * pool.feeThresholdMultiplier
+
+	// The threshold grows exponentially if there are multiple blocks
+	// pending in the pool.
```

Comment:
What's the rationale here behind an exponential growth in threshold? I understand that it prevents spam, but isn't this a relatively aggressive strategy? Why exponential and not, say, quadratic?

Reply:
The benefit of an exponential strategy is that we don't have to pick any magic fee.  With a quadratic approach, it would have to be quadratic relative to some starting point (say, MinTxnFee?), but that seems less satisfactory, and our starting point might not be a good guess for the steady-state operation of the txpool (say, if the txpool is operating in a regime when lots of users want to pay well over MinTxnFee), in which case the threshold would ramp up quite slowly to deal with a DoS attack.